pkg/reconciler/trait/reconciler.go

```diff
@@ -0,0 +1,200 @@
+	return func(r *Reconciler) {
+		r.applicator = a
+	}
+}
+
+// A Reconciler reconciles OAM traits by modifying the the KubernetesApplication
+// that packages their referenced workload.
+type Reconciler struct {
+	client     client.Client
+	newTrait   func() Trait
+	modifier   Modifier
+	applicator resource.Applicator
+
+	log    logging.Logger
+	record event.Recorder
+}
+
+// Kind is an OAM trait kind.
+type Kind schema.GroupVersionKind
+
+// A Trait is a type of OAM trait.
+type Trait interface {
+	resource.Conditioned
+	metav1.Object
+	runtime.Object
+	GetWorkloadRef() corev1.ObjectReference
+}
+
+// NewReconciler returns a Reconciler that reconciles OAM traits by fetching
+// their referenced workload's KubernetesApplication and applying modifications.
+func NewReconciler(m ctrl.Manager, trait Kind, o ...ReconcilerOption) *Reconciler {
+	nt := func() Trait {
+		return resource.MustCreateObject(schema.GroupVersionKind(trait), m.GetScheme()).(Trait)
+	}
+
+	r := &Reconciler{
+		client:     m.GetClient(),
+		newTrait:   nt,
+		modifier:   ModifyFn(noopModifier),
+		applicator: resource.ApplyFn(resource.Apply),
+
+		log:    logging.NewNopLogger(),
+		record: event.NewNopRecorder(),
+	}
+
+	for _, ro := range o {
+		ro(r)
+	}
+
+	return r
+}
+
+// Reconcile an OAM trait type by modifying its referenced workload's
+// KubernetesApplication.
+func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("request", req)
+	log.Debug("Reconciling")
+
+	ctx, cancel := context.WithTimeout(context.Background(), reconcileTimeout)
+	defer cancel()
+
+	trait := r.newTrait()
+	if err := r.client.Get(ctx, req.NamespacedName, trait); err != nil {
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errGetTrait)
+	}
+
+	log = log.WithValues("uid", trait.GetUID(), "version", trait.GetResourceVersion())
+
+	app := &workloadv1alpha1.KubernetesApplication{}
+	err := r.client.Get(ctx, types.NamespacedName{Name: trait.GetWorkloadRef().Name, Namespace: trait.GetNamespace()}, app)
+	if kerrors.IsNotFound(err) {
+		log.Debug("Waiting for referenced workload's KubernetesApplication", "kind", trait.GetObjectKind().GroupVersionKind().String())
+		r.record.Event(trait, event.Normal(reasonTraitWait, "Waiting for workload package to exist"))
+		trait.SetConditions(v1alpha1.ReconcileSuccess())
+		return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, trait), errUpdateTraitStatus)
+	}
+	if err != nil {
+		log.Debug("Cannot get workload package", "error", err, "requeue-after", time.Now().Add(shortWait))
+		r.record.Event(trait, event.Warning(reasonCannotGetPackage, err))
+		trait.SetConditions(v1alpha1.ReconcileError(errors.Wrap(err, errGetPackage)))
+		return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, trait), errUpdateTraitStatus)
+	}
+
+	if err := r.modifier.Modify(ctx, app, trait); err != nil {
```

Comment:
`modifier.Modify` stutters a little. I wonder if there's a name for modifier that would make this read more like English prose?

Reply:
We could rename `modifier` to `trait` in the controller (similar to your suggestion to change `packager` -> `workload`), then we would end up with `r.trait.Modify`